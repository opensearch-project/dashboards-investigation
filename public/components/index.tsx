/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters } from '../../../../src/core/public';
import { NoteBookServices } from '../types';
import { App } from './app';
import { OpenSearchDashboardsContextProvider } from '../../../../src/plugins/opensearch_dashboards_react/public';

export const Observability = (
  services: NoteBookServices,
  AppMountParametersProp: AppMountParameters
) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <App />
    </OpenSearchDashboardsContextProvider>,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
