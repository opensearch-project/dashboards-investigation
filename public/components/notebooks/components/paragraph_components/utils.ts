/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { NotebookType } from '../../../../../common/types/notebooks';

/**
 * Checks if this paragraph is in an agentic notebook and has been run before
 * @param notebookType - The type of notebook
 * @param paragraphIndex - The index of the current paragraph
 * @param totalParagraphs - The total number of paragraphs in the notebook
 * @returns true if the paragraph is agentic and has been run before (not the last paragraph)
 */
export const isAgenticRunBefore = (
  notebookType: NotebookType | undefined,
  paragraphIndex: number,
  totalParagraphs: number
): boolean => {
  return notebookType === NotebookType.AGENTIC && paragraphIndex < totalParagraphs - 1;
};
