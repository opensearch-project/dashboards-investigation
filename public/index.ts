/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './components/notebooks/index.scss';
import { InvestigationPlugin } from './plugin';
import './variables.scss';

export { InvestigationPlugin as Plugin };

export const plugin = () => new InvestigationPlugin();

export { InvestigationStart } from './types';
