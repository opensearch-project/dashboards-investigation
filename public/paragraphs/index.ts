/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AI_RESPONSE_TYPE,
  DASHBOARDS_VISUALIZATION_TYPE,
  DATA_DISTRIBUTION_PARAGRAPH_TYPE,
  DEEP_RESEARCH_PARAGRAPH_TYPE,
  LOG_PATTERN_PARAGRAPH_TYPE,
  OBSERVABILITY_VISUALIZATION_TYPE,
  OTHER_PARAGRAPH_TYPE,
} from '../../common/constants/notebooks';
import { DataDistributionParagraphItem } from './data_distribution';
import { LogPatternParagraphItem } from './log_analytics';
import { MarkdownParagraphItem } from './markdown';
import { OtherParagraphItem } from './other';
import { PERAgentParagraphItem } from './per_agent';
import { PPLParagraphItem } from './ppl';
import { VisualizationParagraphItem } from './visualization';

export const paragraphRegistry = [
  { types: [OTHER_PARAGRAPH_TYPE], item: OtherParagraphItem },
  { types: [DATA_DISTRIBUTION_PARAGRAPH_TYPE], item: DataDistributionParagraphItem },
  { types: [LOG_PATTERN_PARAGRAPH_TYPE], item: LogPatternParagraphItem },
  { types: [DEEP_RESEARCH_PARAGRAPH_TYPE, AI_RESPONSE_TYPE], item: PERAgentParagraphItem },
  {
    types: [
      DASHBOARDS_VISUALIZATION_TYPE.toUpperCase(),
      OBSERVABILITY_VISUALIZATION_TYPE.toUpperCase(),
      DASHBOARDS_VISUALIZATION_TYPE,
      OBSERVABILITY_VISUALIZATION_TYPE,
    ],
    item: VisualizationParagraphItem,
  },
  { types: ['ppl', 'sql'], item: PPLParagraphItem },
  { types: ['md'], item: MarkdownParagraphItem },
];
