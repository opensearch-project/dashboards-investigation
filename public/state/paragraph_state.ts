/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ParagraphBackendType } from '../../common/types/notebooks';
import { ObservableState } from './observable_state';

export interface ParagraphStateValue<TFullfilledOutput = {}> extends ParagraphBackendType {
  fullfilledOutput?: Partial<TFullfilledOutput>; // this is the fullfilled output, like PPL query result / PER agent response
  uiState: {
    viewMode: 'input_only' | 'output_only' | 'view_both';
    inQueue?: boolean;
    isRunning?: boolean;
    isOutputStale?: boolean;
  };
}

export class ParagraphState<TFullfilledOutput = {}> extends ObservableState<
  ParagraphStateValue<TFullfilledOutput>
> {
  getParagraphType() {
    return this.value.input.inputType;
  }
  getBackgroundValue() {
    const { input, output, id, dateModified, dateCreated } = this.value;
    return {
      input,
      output,
      id,
      dateModified,
      dateCreated,
    };
  }
  updateInput(input: Partial<ParagraphStateValue['input']>) {
    this.updateValue({
      input: {
        ...this.value.input,
        ...input,
      },
    });
    return this;
  }
  updateOutput(output: Partial<Required<ParagraphStateValue>['output'][0]>) {
    this.updateValue({
      output: [
        {
          ...(this.value.output?.[0] || {}),
          ...output,
        } as Required<ParagraphBackendType>['output'][0],
      ],
    });
  }
  updateFullfilledOutput(fullfilledOutput: Partial<TFullfilledOutput>) {
    this.updateValue({
      fullfilledOutput: {
        ...this.value.fullfilledOutput,
        ...fullfilledOutput,
      },
    });
  }
  updateUIState(uiState: Partial<ParagraphStateValue['uiState']>) {
    this.updateValue({
      uiState: {
        ...this.value.uiState,
        ...uiState,
      },
    });
  }
}
