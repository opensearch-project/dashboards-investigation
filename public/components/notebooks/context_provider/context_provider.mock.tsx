/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useRef } from 'react';
import {
  NotebookContextProvider,
  NotebookContextProviderProps,
  getDefaultState,
} from './context_provider';
import { ParagraphState, ParagraphStateValue } from '../../../../common/state/paragraph_state';

export const MockContextProvider = (
  props: Omit<NotebookContextProviderProps, 'state'> & {
    paragraphValues: ParagraphStateValue[];
  }
) => {
  const defaultStateRef = useRef(
    getDefaultState({
      paragraphs: props.paragraphValues.map((paragraph) => new ParagraphState<unknown>(paragraph)),
    })
  );
  return <NotebookContextProvider {...props} state={defaultStateRef.current} />;
};
