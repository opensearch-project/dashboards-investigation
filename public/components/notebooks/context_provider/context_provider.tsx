/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useReducer } from 'react';
import { NotebookState } from '../../../state/notebook_state';
import { TopContextState } from '../../../state/top_context_state';
import { notebookReducer, Action } from '../reducers/notebook_reducer';
import { HttpStart } from '../../../../../../src/core/public';

const defaultState = new NotebookState({
  paragraphs: [],
  id: '',
  context: new TopContextState({}),
  dataSourceEnabled: false,
  dateCreated: '',
  isLoading: false,
  path: '',
});

export const NotebookReactContext = React.createContext<{
  state: NotebookState;
  dispatch: React.Dispatch<Action>;
  http: HttpStart;
}>({
  state: defaultState,
  dispatch: () => {},
  http: {} as HttpStart,
});

export const NotebookContextProvider = (props: {
  children: React.ReactChild;
  notebookId: string;
  http: HttpStart;
  dataSourceEnabled: boolean;
}) => {
  const [state, dispatch] = useReducer(
    notebookReducer,
    new NotebookState({
      paragraphs: [],
      id: props.notebookId,
      context: new TopContextState({}),
      dataSourceEnabled: props.dataSourceEnabled,
      isLoading: true,
      dateCreated: '',
      path: '',
    })
  );

  return (
    <NotebookReactContext.Provider
      value={{
        state,
        dispatch,
        http: props.http,
      }}
    >
      {props.children}
    </NotebookReactContext.Provider>
  );
};
