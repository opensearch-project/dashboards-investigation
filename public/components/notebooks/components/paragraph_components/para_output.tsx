/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCodeBlock, EuiSpacer, EuiText } from '@elastic/eui';
import MarkdownRender from '@nteract/markdown';
import { Media } from '@nteract/outputs';
import moment from 'moment';
import React from 'react';
import { useContext } from 'react';
import { Observable } from 'rxjs';
import { ParagraphStateValue } from 'common/state/paragraph_state';
import { LogPatternAnalysisResult } from 'common/types/log_pattern';
import { NoteBookServices } from 'public/types';
import {
  ANOMALY_VISUALIZATION_ANALYSIS_PARAGRAPH_TYPE,
  LOG_PATTERN_PARAGRAPH_TYPE,
} from '../../../../../common/constants/notebooks';
import { DashboardContainerInput } from '../../../../../../../src/plugins/dashboard/public';
import {
  AnomalyVisualizationAnalysisOutputResult,
  ParaType,
} from '../../../../../common/types/notebooks';
import { uiSettingsService } from '../../../../../common/utils';
import { QueryDataGridMemo } from './para_query_grid';
import { BubbleUpContainer } from '../bubbleup/bubble_up_container';
import { LogPatternContainer } from '../log_analytics/log_pattern_container';
import { DashboardPanelState } from '../../../../../../../src/plugins/dashboard/public/application';
import { EmbeddableInput } from '../../../../../../../src/plugins/embeddable/public';
import { NotebookReactContext } from '../../context_provider/context_provider';
import { useOpenSearchDashboards } from '../../../../../../../src/plugins/opensearch_dashboards_react/public';

const createQueryColumns = (jsonColumns: any[]) => {
  let index = 0;
  const datagridColumns = [];
  for (index = 0; index < jsonColumns.length; ++index) {
    const datagridColumnObject = {
      id: jsonColumns[index].name,
      displayAsText: jsonColumns[index].name,
    };
    datagridColumns.push(datagridColumnObject);
  }
  return datagridColumns;
};

const getQueryOutputData = (queryObject: any) => {
  const data = [];
  let index = 0;
  let schemaIndex = 0;
  for (index = 0; index < queryObject.datarows.length; ++index) {
    const datarowValue = {};
    for (schemaIndex = 0; schemaIndex < queryObject.schema.length; ++schemaIndex) {
      const columnName = queryObject.schema[schemaIndex].name;
      if (typeof queryObject.datarows[index][schemaIndex] === 'object') {
        datarowValue[columnName] = JSON.stringify(queryObject.datarows[index][schemaIndex]);
      } else if (typeof queryObject.datarows[index][schemaIndex] === 'boolean') {
        datarowValue[columnName] = queryObject.datarows[index][schemaIndex].toString();
      } else {
        datarowValue[columnName] = queryObject.datarows[index][schemaIndex];
      }
    }
    data.push(datarowValue);
  }
  return data;
};

const OutputBody = ({
  index,
  typeOut,
  val,
  para,
  visInput,
  setVisInput,
}: {
  index: number;
  typeOut: string;
  val: string;
  para: ParaType;
  visInput: DashboardContainerInput;
  setVisInput: (input: DashboardContainerInput) => void;
}) => {
  /* Returns a component to render paragraph outputs using the para.typeOut property
   * Currently supports HTML, TABLE, IMG
   * TODO: add table rendering
   */
  const context = useContext(NotebookReactContext);
  const {
    services: {
      dashboard: { DashboardContainerByValueRenderer },
    },
  } = useOpenSearchDashboards<NoteBookServices>();

  const paragraph$: Observable<ParagraphStateValue<any>> = context.state.value.paragraphs[
    index
  ].getValue$();

  const dateFormat = uiSettingsService.get('dateFormat');

  if (typeOut !== undefined) {
    switch (typeOut) {
      case 'QUERY':
        const inputQuery = para.inp.substring(4, para.inp.length);
        const queryObject = JSON.parse(val);
        if (queryObject.hasOwnProperty('error')) {
          return <EuiCodeBlock>{val}</EuiCodeBlock>;
        } else {
          const columns = createQueryColumns(queryObject.schema);
          const data = getQueryOutputData(queryObject);
          return (
            <div>
              <EuiText key={'query-input-key'} className="wrapAll" data-test-subj="queryOutputText">
                <b>{inputQuery}</b>
              </EuiText>
              <EuiSpacer />
              <QueryDataGridMemo
                rowCount={queryObject.datarows.length}
                queryColumns={columns}
                dataValues={data}
              />
            </div>
          );
        }
      case 'MARKDOWN':
        return (
          <EuiText
            className="wrapAll markdown-output-text"
            data-test-subj="markdownOutputText"
            size="s"
          >
            <MarkdownRender source={val} />
          </EuiText>
        );
      case 'VISUALIZATION':
      case 'OBSERVABILITY_VISUALIZATION':
        let from = moment(visInput?.timeRange?.from).format(dateFormat);
        let to = moment(visInput?.timeRange?.to).format(dateFormat);
        from = from === 'Invalid date' ? visInput.timeRange.from : from;
        to = to === 'Invalid date' ? visInput.timeRange.to : to;

        const panels = Object.entries(visInput?.panels || {}).reduce(
          (
            acc,
            [panelKey, panel]: [
              string,
              DashboardPanelState<EmbeddableInput & { [k: string]: unknown }>
            ]
          ) => {
            const savedObjectId: string = panel.explicitInput.savedObjectId as string;
            if (savedObjectId.includes('observability-visualization:')) {
              return {
                ...acc,
                [panelKey]: {
                  ...panel,
                  type: 'observability-visualization',
                  explicitInput: {
                    ...panel.explicitInput,
                    savedObjectId: savedObjectId.replace('observability-visualization:', ''),
                  },
                },
              };
            }

            return {
              ...acc,
              [panelKey]: panel,
            };
          },
          {} as DashboardContainerInput['panels']
        );

        return (
          <>
            <EuiText size="s" style={{ marginLeft: 9 }}>
              {`${from} - ${to}`}
            </EuiText>
            <DashboardContainerByValueRenderer
              input={{
                ...visInput,
                panels,
              }}
              onInputUpdated={setVisInput}
            />
          </>
        );
      case 'HTML':
        return (
          <EuiText>
            {/* eslint-disable-next-line react/jsx-pascal-case */}
            <Media.HTML data={val} />
          </EuiText>
        );
      case 'TABLE':
        return <pre>{val}</pre>;
      case 'IMG':
        return <img alt="" src={'data:image/gif;base64,' + val} />;
      case ANOMALY_VISUALIZATION_ANALYSIS_PARAGRAPH_TYPE:
        return (
          <BubbleUpContainer
            paragraph$={
              paragraph$ as Observable<
                ParagraphStateValue<AnomalyVisualizationAnalysisOutputResult>
              >
            }
          />
        );
      case LOG_PATTERN_PARAGRAPH_TYPE:
        return (
          <LogPatternContainer
            para={para}
            paragraph$={paragraph$ as Observable<ParagraphStateValue<LogPatternAnalysisResult>>}
          />
        );
      default:
        return <pre>{val}</pre>;
    }
  } else {
    console.log('output not supported', typeOut);
    return <pre />;
  }
};

/*
 * "ParaOutput" component is used by notebook to populate paragraph outputs for an open notebook.
 *
 * Props taken in as params are:
 * para - parsed paragraph from notebook
 *
 * Outputs component of nteract used as a container for notebook UI.
 * https://components.nteract.io/#outputs
 */
export interface ParaOutputProps {
  index: number;
  para: ParaType;
  visInput: DashboardContainerInput;
  setVisInput: (input: DashboardContainerInput) => void;
}

export const ParaOutput = (props: ParaOutputProps) => {
  const { index, para, visInput, setVisInput } = props;

  return (
    !para.isRunning && (
      <>
        {para.typeOut.map((typeOut: string, tIdx: number) => {
          return (
            <OutputBody
              index={index}
              key={para.uniqueId + '_paraOutputBody'}
              typeOut={typeOut}
              val={para.out[tIdx]}
              para={para}
              visInput={visInput}
              setVisInput={setVisInput}
            />
          );
        })}
      </>
    )
  );
};
