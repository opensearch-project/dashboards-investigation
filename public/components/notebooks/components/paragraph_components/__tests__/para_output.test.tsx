/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { uiSettingsService } from '../../../../../../common/utils/core_services';
import {
  sampleObservabilityVizParagraph,
  sampleParsedParagraghs1,
} from '../../../../../../test/notebooks_constants';
import { ParaOutput, ParaOutputProps } from '../para_output';
import { ParagraphStateValue } from 'common/state/paragraph_state';
import { MockContextProvider } from '../../../context_provider/context_provider.mock';
import { OpenSearchDashboardsContextProvider } from '../../../../../../../../src/plugins/opensearch_dashboards_react/public';

jest.mock('../../bubbleup/bubble_up_container', () => ({
  BubbleUpContainer: () => <div />,
}));

const ContextAwareParaOutput = (
  props: ParaOutputProps & {
    paragraphValues: ParagraphStateValue[];
  }
) => {
  return (
    <OpenSearchDashboardsContextProvider
      services={{ dashboard: { DashboardContainerByValueRenderer: () => null } }}
    >
      <MockContextProvider paragraphValues={props.paragraphValues}>
        <ParaOutput {...props} />
      </MockContextProvider>
    </OpenSearchDashboardsContextProvider>
  );
};

describe('<ParaOutput /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders markdown outputs', () => {
    const para = sampleParsedParagraghs1[0];
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders query outputs', () => {
    const para = sampleParsedParagraghs1[3];
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders query outputs with error', () => {
    const para = sampleParsedParagraghs1[3];
    para.out = ['{"error":"Invalid SQL query"}'];
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders dashboards visualization outputs', () => {
    const para = sampleParsedParagraghs1[2];

    uiSettingsService.get = jest.fn().mockReturnValue('YYYY-MMM-DD HH:mm:ss');
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={{
          timeRange: { from: '2020-JUL-21 18:37:44', to: '2020-AUG-20 18:37:44' },
        }}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.textContent).toMatch('2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44');
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders observability visualization outputs', () => {
    const para = sampleObservabilityVizParagraph;

    uiSettingsService.get = jest.fn().mockReturnValue('YYYY-MMM-DD HH:mm:ss');
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={{
          timeRange: { from: '2020-JUL-21 18:37:44', to: '2020-AUG-20 18:37:44' },
        }}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.textContent).toMatch('2020-Jul-21 18:37:44 - 2020-Aug-20 18:37:44');
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders other types of outputs', () => {
    const para = sampleParsedParagraghs1[0];
    para.typeOut = ['HTML', 'TABLE', 'IMG', 'UNKNOWN', undefined];
    para.out = ['', '', '', '', ''];
    const setVisInput = jest.fn();
    const utils = render(
      <ContextAwareParaOutput
        key={para.uniqueId}
        para={para}
        visInput={jest.fn()}
        setVisInput={setVisInput}
        paragraphValues={[para]}
        index={0}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
