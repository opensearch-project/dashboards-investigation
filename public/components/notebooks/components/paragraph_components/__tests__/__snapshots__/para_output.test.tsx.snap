// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParaOutput /> spec renders markdown outputs 1`] = `
<pre>
  # Type your input here
</pre>
`;

exports[`<ParaOutput /> spec renders other types of outputs 1`] = `
<div
  class="euiText euiText--medium"
>
  <div
    class="sc-eDDNvR gUOHUY"
  />
</div>
`;

exports[`<ParaOutput /> spec renders query outputs 1`] = `
<pre>
  {"schema":[{"name":"FlightNum","type":"keyword"},{"name":"Origin","type":"keyword"},{"name":"OriginLocation","type":"geo_point"},{"name":"DestLocation","type":"geo_point"},{"name":"FlightDelay","type":"boolean"},{"name":"DistanceMiles","type":"float"},{"name":"FlightTimeMin","type":"float"},{"name":"OriginWeather","type":"keyword"},{"name":"dayOfWeek","type":"integer"},{"name":"AvgTicketPrice","type":"float"},{"name":"Carrier","type":"keyword"},{"name":"FlightDelayMin","type":"integer"},{"name":"OriginRegion","type":"keyword"},{"name":"DestAirportID","type":"keyword"},{"name":"FlightDelayType","type":"keyword"},{"name":"timestamp","type":"timestamp"},{"name":"Dest","type":"keyword"},{"name":"FlightTimeHour","type":"keyword"},{"name":"Cancelled","type":"boolean"},{"name":"DistanceKilometers","type":"float"},{"name":"OriginCityName","type":"keyword"},{"name":"DestWeather","type":"keyword"},{"name":"OriginCountry","type":"keyword"},{"name":"DestCountry","type":"keyword"},{"name":"DestRegion","type":"keyword"},{"name":"DestCityName","type":"keyword"},{"name":"OriginAirportID","type":"keyword"}],"datarows":[["9HY9SWR","Frankfurt am Main Airport",{"lat":50.033333,"lon":8.570556},{"lat":-33.94609833,"lon":151.177002},false,10247.856,1030.7704,"Sunny",0,841.2656,"OpenSearch Dashboards Airlines",0,"DE-HE","SYD","No Delay","2021-11-01 00:00:00","Sydney Kingsford Smith International Airport","17.179506930998397",false,16492.326,"Frankfurt am Main","Rain","DE","AU","SE-BD","Sydney","FRA"],["X98CCZO","Cape Town International Airport",{"lat":-33.96480179,"lon":18.60169983},{"lat":45.505299,"lon":12.3519},false,5482.6064,464.3895,"Clear",0,882.98267,"Logstash Airways",0,"SE-BD","VE05","No Delay","2021-11-01 18:27:00","Venice Marco Polo Airport","7.73982468459836",false,8823.4,"Cape Town","Sunny","ZA","IT","IT-34","Venice","CPT"]],"total":2,"size":2,"status":200}
</pre>
`;

exports[`<ParaOutput /> spec renders query outputs with error 1`] = `
<pre>
  {"error":"Invalid SQL query"}
</pre>
`;
